### PR TITLE
New Add & Edit Command and Person Fields 

### PR DESCRIPTION
**Pull request by Charles1026**
image

## `add` and `edit` Command
The `add` and `edit` command argument fields have been updated to support the required fields for patients as shown below

<table>
<tr>
<th> Old </th><th> New </th>
</tr>
<tr>
<td> Name (n/) </td><td> Name (n/) </td>
</tr>
<tr>
<td> Phone (p/) </td><td> Phone (p/) </td>
</tr>
<tr>
<td> Email (e/) </td><td> Email (e/) </td>
</tr>
<tr>
<td> address (a/) </td><td> Next Of Kin (nok/) </td>
</tr>
<tr>
<td> NIL </td><td> Patient Type (pt/) </td>
</tr>
<tr>
<td> NIL </td><td> Hospital Wing (hw/) </td>
</tr>
<tr>
<td> NIL </td><td> Floor Number (fn/) </td>
</tr>
<tr>
<td> NIL </td><td> Ward Number (wn/) </td>
</tr>
<tr>
<td> NIL </td><td> Medication (m/) </td>
</tr>
</table>

### Parsing Support for Inpatient/Outpatient
As it does not make sense for an outpatient to possess wing, floor, and ward attributes. Parsing support has been added to ensure that when the patient is set to an outpatient, none of the three fields will be accepted and when the patient is an inpatient, all of the three fields are required.

## `Person` Fields
Similar to the `add` command, `Person` has been updated to store and display(as seen in the image above) the new attributes added through the creation of new attribute classes as shown below:

### `NextOfKin` Class
NOK information is input and stored in this class as a string in the format `name, relationship, contact number`. Regex support has been added to ensure the input conforms to this format.

### `PatientType` Class
This class stores information on whether a patient is an inpatient or an outpatient. Input can only be of the form `"i"`, `"inpatient"`, `"o"` or `"outpatient"` and is case insensitive. The information is stored as an enum within the class itself.
To facilitate easy discerning the method `public bool isInpatient()` is included in the class.

### `HospitalWing` Class
This class's regex validation has been updated to only allow alphanumeric characters and spaces and must not be empty if present.
* Note that this is a field stored as an `Optional` in the `Person` class.

### `FloorNumber` & `WardNumber` Class
These 2 classes store the patient's floor and ward numbers respectively. The input string is parsed in `ParserUtil` into an `Integer` which the class takes in. Then, validation is carried out to ensure that the input is in fact a positive Integer.
* Note that this is a field stored as an `Optional` in the `Person` class.

### `Medication` Class
This class was converted from the `Tag` class and functions identically. Regex support was added to ensure the values are alphanumeric and space characters only.
* Note that this is a field stored as an `Optional` in the `Person` class.

##ToDo
- Beautify the UI, currently, the UI support for these new fields is very basic and looks terrible, color and proper margins can be added to make it look better.
- Standardise `PatientType`, `FloorNumber`, and `WardNumber` to store their values as `String` instead of their own types to ensure consistency across the board for all fields. Proper regex validation will be needed to facilitate this.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/NextOfKin.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Represents a Patient's Next of Kin in the Patient Database.
+ * Guarantees: immutable; is valid as declared in {@link #isValidNextOfKin(String)}
+ */
+public class NextOfKin {
+
+    public static final String MESSAGE_CONSTRAINTS = "Next of Kin should not be blank "
+            + "and should come in the format: Name, Relationship, Contact";
+
+    public static final String VALIDATION_REGEX = "[A-Za-z\\s]+,[a-zA-Z\\s]+,[0-9\\s]+";
```
Comment (PeiYee88) [inline src/main/java/seedu/address/model/person/NextOfKin.java:15]: Validation regex looks good! 
 ** For parser regex for get /nok, will need to consider the case when name has spaces in between (First and last name).

**Comment by nehcuy** [comment]
Great job implementing add, thanks for the effort! I think we can merge this.

**Review by PeiYee88 — APPROVED** [review]
No major issues found atm!

File: src/main/java/seedu/address/model/person/Person.java [hunk 2]
```diff
@@ -21,21 +22,29 @@ public class Person {
...
 
     // Data fields
-    private final Address address;
-    private final HospitalWing hospitalWing;
-    private final Set<Tag> tags = new HashSet<>();
+    private final NextOfKin nextOfKin;
+    private final PatientType patientType;
+    private final Optional<HospitalWing> hospitalWing;
+    private final Optional<FloorNumber> floorNumber;
+    private final Optional<WardNumber> wardNumber;
```
Comment (nehcuy) [inline src/main/java/seedu/address/model/person/Person.java:29]: Great use of `Optional` for the case of outpatients not having these displayed in their `PersonCard`!

Comment (nehcuy) [reply src/main/java/seedu/address/model/person/Person.java:29]: Great job implementing add, thanks for the effort! I think we can merge this.

File: docs/UserGuide.md [hunk 3]
```diff
@@ -151,7 +151,7 @@ Finds total number of patients.
 Format: `count`
 
 * A count of all existing patients in the hospital will be returned.
-* The count will be a non-negative integer (>= 0).  
+* The count will be a non-negative integer (>= 0).
```
Comment (kxrt) [inline docs/UserGuide.md:154]: Thanks for fixing all the whitespaces!

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 4]
```diff
@@ -24,17 +27,23 @@ public class AddCommand extends Command {
...
-            + PREFIX_ADDRESS + "311, Clementi Ave 2, #02-25 "
+            + PREFIX_NEXT_OF_KIN + "Jane Doe, Wife, 82858285 "
+            + PREFIX_PATIENT_TYPE + "inpatient "
             + PREFIX_HOSPITAL_WING + "south "
-            + PREFIX_TAG + "friends "
-            + PREFIX_TAG + "owesMoney";
+            + PREFIX_FLOOR_NUMBER + "3 "
+            + PREFIX_WARD_NUMBER + "69 "
+            + PREFIX_MEDICATION + "panadol "
+            + PREFIX_MEDICATION + "ibuprofen ";
```
Comment (kxrt) [inline src/main/java/seedu/address/logic/commands/AddCommand.java:46]: Very good variable naming

File: src/main/java/seedu/address/logic/parser/Prefix.java [hunk 5]
```diff
@@ -15,6 +15,7 @@ public String getPrefix() {
         return prefix;
     }
 
+    @Override
```
Comment (kxrt) [inline src/main/java/seedu/address/logic/parser/Prefix.java:18]: Nice catch!

File: src/main/java/seedu/address/model/person/WardNumber.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * Constructs an {@code Ward Number}.
+     *
+     * @param wardNumber A valid ward number.
+     */
+    public WardNumber(Integer wardNumber) {
+        checkArgument(isValidWardNumber(wardNumber), MESSAGE_CONSTRAINTS);
+        value = wardNumber;
+    }
```
Comment (kxrt) [inline src/main/java/seedu/address/model/person/WardNumber.java:23]: Using `int` over the wrapper type `Integer` might help with object housekeeping, but that can be ignored for now.

**Review by kxrt — APPROVED** [review]
Lgtm, very carefully edited the code structure!

File: src/main/java/seedu/address/model/person/WardNumber.java (hunk 6 above)
Comment (Charles1026) [reply src/main/java/seedu/address/model/person/WardNumber.java:23]: Agree, was considering using int, will change it in future or just use string with regex

File: src/main/java/seedu/address/model/person/NextOfKin.java (hunk 1 above)
Comment (Charles1026) [reply src/main/java/seedu/address/model/person/NextOfKin.java:15]: Agree, for now, it only checks that the three fields are there, no validation is done for the actual fields, only that they must have at least 1 alpha/numeric or space character.

**Pull request merged**